### ISSUE TITLE
☂️  Prettier Divergence collection from the Prettier challenge

### ISSUE THREAD

**Issue by faultyserver** [biomejs/biome#739]
_**NOTE:** This issue thread is for _explaining intentional divergences_ from Prettier. It is not intended as a place to seek clarification about formatting rules or point out discrepancies with Prettier. For those, please open a new issue or discussion. Once a decision has been made, a new entry will be added to this thread explaining the rationale and purpose behind choosing to diverge from Prettier's rules._

This Issue will function as a hub for collecting divergence from Prettier. When you want to comment, please use the following template:

```markdown
## Divergence

<! -- Add a diff and the name of the file from `report.md` -->

## Reason

<! -- Reason why we should diverge -->
```

**Comment by ematipico** [comment]
## Divergence

Input
```typescript
// Multiple accessibility modifiers
class Foo {
  private public a  = 1;
}

// Declare function with body
declare function foo ( ) {  }

// Invalid use of abstract
class Bar {
  abstract  foo  ;
}

// Duplicate Readonly
class Read {
  readonly readonly   x: number;
}
```

Diff
```diff
// Multiple accessibility modifiers
class Foo {
-  private a = 1;
+  private public a  = 1;
}

// Declare function with body
-declare function foo() {};
+declare function foo ( ) {  }

// Invalid use of abstract
class Bar {
-  abstract foo;
+  abstract  foo  ;
}

// Duplicate Readonly
class Read {
-  readonly x: number;
+  readonly readonly   x: number;
}
```

### Related Prettier Tests

- [typescript/conformance/classes/classDeclarations/classAbstractKeyword/classAbstractProperties.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report.md#typescriptconformanceclassesclassdeclarationsclassabstractkeywordclassabstractpropertiests)
- [typescript/declare/declare_function_with_body.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report.md#typescriptdeclaredeclare_function_with_bodyts)
- [typescript/conformance/classes/constructorDeclarations/constructorParameters/readonlyReadonly.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report.md#typescriptconformanceclassesconstructordeclarationsconstructorparametersreadonlyreadonlyts)
- [typescript/conformance/classes/classDeclarations/classAbstractKeyword/classAbstractInstantiations2.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report.md#typescriptconformanceclassesclassdeclarationsclassabstractkeywordclassabstractinstantiations2ts)
- [typescript/conformance/classes/classDeclarations/classAbstractKeyword/classAbstractMixedWithModifiers.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report.md#typescriptconformanceclassesclassdeclarationsclassabstractkeywordclassabstractmixedwithmodifiersts)
- [typescript/custom/abstract/abstractProperties.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report.md#typescriptcustomabstractabstractpropertiests)

## Reason

Prettier's Babel-based parsing for JS/TS is very loose and [allows multiple errors](https://github.com/prettier/prettier/blob/e4a74c05f4502dd4ec70495c3130ff08ab088e05/src/language-js/parse/babel.js#L177-L218) to be ignored. Biome's parser intentionally has stricter rules than Prettier and correctly asserts that the statements in these tests are invalid. Duplicate modifiers for functions are not semantically valid, function declarations are not allowed to have bodies, non-abstract classes cannot have abstract properties, etc.

In Prettier, these errors aren't considered parse errors, and the AST is still built "correctly" with the appropriate nodes. When formatting, Prettier then just treats these nodes as normal and formats them accordingly.

In Biome, the parsing errors result in `Bogus` nodes, which may contain any number of valid nodes, invalid nodes, and/or raw tokens. When formatting, Biome treats bogus nodes as effectively plain text, printing them out verbatim into the resulting code without any formatting, since attempting to format could be incorrect and cause semantic changes.

For class properties, Prettier's current parsing strategy also uses boolean fields for modifiers, meaning only one of each kind of modifier can ever be present (accessibility modifiers are stored as a single string). When printing, Prettier just looks at the list of booleans and decides which modifiers to print out again. Biome instead keeps a list of modifiers, meaning duplicates are kept around and able to be analyzed (hence the parsing error messages about duplicate modifiers and ordering). When printing out the bogus nodes, this list is still kept intact, and printing out the unformatted text results in those modifiers continuing to exist.

There are ways that Biome can address this. One possibility is to try to interpret the Bogus nodes when formatting and construct valid nodes out of them. If a valid node can be built, then it would just format that node like normal, otherwise it prints the bogus text verbatim as it does currently. However, this is messy and introduces a form of parsing logic into the formatter that is really not meaningful.

Another option instead is to introduce some form of "syntactically-valid bogus node" into the parser, which accepts these kinds of errors that are purely semantic (duplicate modifiers, abstract properties in non-abstract classes). It would continue to build the nodes like normal (effectively matching the behavior in Prettier), but store them inside of a the new kind of bogus node, including the diagnostics along with it. When formatting, these special bogus nodes would just attempt to format the inner node, and then fallback if there's an error (the existing `format_or_verbatim` utility would do this already). This keeps the parsing and formatting logic separate from each other, but introduce more complexity to the parser, allowing invalid states to be considered semi-valid.

After [a little bit of discussion in Discord](https://discord.com/channels/1132231889290285117/1132602615570649110/1174492440275722330), the decision for now is to not worry about this divergence unless it's needed to reach 95% compatibility with Prettier.

## Playground link

https://biomejs.dev/playground/?indentStyle=space&quoteStyle=single&trailingComma=none&code=LwAvACAATQB1AGwAdABpAHAAbABlACAAYQBjAGMAZQBzAHMAaQBiAGkAbABpAHQAeQAgAG0AbwBkAGkAZgBpAGUAcgBzAAoAYwBsAGEAcwBzACAARgBvAG8AIAB7AAoAIAAgAHAAcgBpAHYAYQB0AGUAIABwAHUAYgBsAGkAYwAgAGEAIAAgAD0AIAAxADsACgB9AAoACgAvAC8AIABEAGUAYwBsAGEAcgBlACAAZgB1AG4AYwB0AGkAbwBuACAAdwBpAHQAaAAgAGIAbwBkAHkACgBkAGUAYwBsAGEAcgBlACAAZgB1AG4AYwB0AGkAbwBuACAAZgBvAG8AIAAoACAAKQAgAHsAIAAgAH0ACgAKAC8ALwAgAEkAbgB2AGEAbABpAGQAIAB1AHMAZQAgAG8AZgAgAGEAYgBzAHQAcgBhAGMAdAAKAGMAbABhAHMAcwAgAEIAYQByACAAewAKACAAIABhAGIAcwB0AHIAYQBjAHQAIAAgAGYAbwBvACAAIAA7AAoAfQAKAAoALwAvACAARAB1AHAAbABpAGMAYQB0AGUAIABSAGUAYQBkAG8AbgBsAHkACgBjAGwAYQBzAHMAIABSAGUAYQBkACAAewAKACAAIAByAGUAYQBkAG8AbgBsAHkAIAByAGUAYQBkAG8AbgBsAHkAIAAgACAAeAA6ACAAbgB1AG0AYgBlAHIAOwAKAH0A

**Comment by Conaclos** [comment]
## Divergence

Prettier doesn't unquote some object properties that are valid JavaScript identifiers.

Input

```js
const obj = {
 'a': true,
 b: true,
 "𐊧": true,
}
```

Diff

```diff
 const obj = {
   a: true,
   b: true,
-  "𐊧": true,
+  𐊧: true,
 };
```

[Playground link](https://biomejs.dev/playground/?indentStyle=space&quoteStyle=single&trailingComma=none&code=YwBvAG4AcwB0ACAAbwBiAGoAIAA9ACAAewAKACAAJwBhACcAOgAgAHQAcgB1AGUALAAKACAAYgA6ACAAdAByAHUAZQAsAAoAIAAiAADYp94iADoAIAB0AHIAdQBlACwACgB9AA%3D%3D)

Prettier test

- [js/quotes/objects.js](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report.md#jsquotesobjectsjs)

## Reason

Prettier and Biome unquote object and class properties that are valid JavaScript identifiers.
Prettier [unquotes only valid ES5 identifiers](https://github.com/prettier/prettier/blob/a5d502513e5de4819a41fd90b9be7247146effc7/src/language-js/utils/index.js#L646).

In an ecosystem where ES2015 is now widespread, this seems to be a legacy restriction.
Thus, we decided to diverge here by unquoting all valid JavaScript identifiers in ES2015+.

A possible workaround would be to introduce a configuration to set the ECMAScript version a project uses.
We could then adjust the unquoting behaviour based on that version.
Setting the ECMAScript version to `ES5` could match the behaviour of Prettier.

**Comment by Conaclos** [comment]
## Divergence

Input

```js
a = {
  [this.resource = resource]: 1,
}
```

Diff

```diff
 a = {
-  [(this.resource = resource)]: 1,
+  [this.resource = resource]: 1,
 };
```

[Playground link](https://biomejs.dev/playground?code=YQAgAD0AIAB7AAoAIAAgAFsAdABoAGkAcwAuAHIAZQBzAG8AdQByAGMAZQAgAD0AIAByAGUAcwBvAHUAcgBjAGUAXQA6ACAAMQAsAAoAfQAKAA%3D%3D)

Prettier Test

- [js/objects/assignment-expression/object-property.js](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report.md#jsobjectsassignment-expressionobject-propertyjs)

## Reason

Prettier and Biome enclose some assignment expressions between parentheses, in particular in conditionals.
This allows to identify potentially an expression that should be a comparison.

Prettier has an inconsistent behavior, because it adds parentheses for an assignment in a computed key of an object property and doesn't for a computed key of a class property, as demonstrated by the following example:

Input

```js
a = {
  [x = 0]: 1,
}

class C {
  [x = 0] = 1
}
```

Output

```js
a = {
  [(x = 0)]: 1,
};

class C {
  [x = 0] = 1;
}
```

[Playground link](https://biomejs.dev/playground?enabledLinting=false&code=YQAgAD0AIAB7AAoAIAAgAFsAeAAgAD0AIAAwAF0AOgAgADEALAAKAH0ACgAKAGMAbABhAHMAcwAgAEMAIAB7AAoAIAAgACAAIABbAHgAIAA9ACAAMABdACAAPQAgADEACgB9AAoA)

To be consistent, we decide to diverge and omit the parentheses.
Alternatively, we could enclose any assignment in a computed key of an object or of a class.

**Comment by Conaclos** [comment]
## Divergence

Input

```js
interface L<in const T> {}
```

Diff

```diff
-interface L<const in T> {}
+interface L<in const T> {}
```

[Playground link](https://biomejs.dev/playground?enabledLinting=false&code=aQBuAHQAZQByAGYAYQBjAGUAIABMADwAaQBuACAAYwBvAG4AcwB0ACAAVAA%2BACAAewB9AA%3D%3D)

Prettier Test

- [typescript/typeparams/const.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report.md#typescripttypeparamsconstts)

## Reason

As mentioned in a [previous divergence](https://github.com/biomejs/biome/issues/739#issuecomment-1815374273), Prettier's Babel-based parsing for JS/TS is very loose and allows multiple errors to be ignored. Biome's parser intentionally has stricter rules than Prettier and correctly asserts that the `const` modifier is not allowed for type parameters of interfaces.

In Prettier, this error isn't considered a parse error, and the AST is still built "correctly" with the appropriate nodes. When formatting, Prettier then just treats these nodes as normal and formats them accordingly.

In Biome, the parsing errors result in a Bogus node, which may contain any number of valid nodes, invalid nodes, and/or raw tokens. When formatting, Biome treats bogus nodes as effectively plain text, printing them out verbatim into the resulting code without any formatting, since attempting to format could be incorrect and cause semantic changes.

We could adopt the same strategy that discussed in a [previous comment](https://github.com/biomejs/biome/issues/739#issuecomment-1815374273) to fix the divergence.

**Comment by Conaclos** [comment]
## Divergence

Prettier encloses optional chains between parentheses.

Input

```js
(a?.b).c;
(a?.()).b;

(a?.b)();
(a?.())();

new (a?.b)();
new (a?.())();
```

Diff

```diff
-(a?.b).c;
-(a?.()).b;
+a?.b.c;
+a?.().b;
 
-(a?.b)();
-(a?.())();
+a?.b();
+a?.()();
 
 new (a?.b)();
 new (a?.())();
```

[Playground link](https://biomejs.dev/playground?enabledLinting=false&code=KABhAD8ALgBiACkALgBjADsACgAoAGEAPwAuACgAKQApAC4AYgA7AAoACgAoAGEAPwAuAGIAKQAoACkAOwAKACgAYQA%2FAC4AKAApACkAKAApADsACgAKAG4AZQB3ACAAKABhAD8ALgBiACkAKAApADsACgBuAGUAdwAgACgAYQA%2FAC4AKAApACkAKAApADsA)

PrettierTests

- [js/chain-expression/test.js](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#jschain-expressiontestjs)
- [js/optional-chaining-assignment/valid-parenthesized.js](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report.md#jsoptional-chaining-assignmentvalid-parenthesizedjs)
- [typescript/chain-expression/test.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report.md#typescripttypeparamsconstts)

## Reason

Prettier encloses every optional chain between parentheses.
As [discussed on Discord](https://discord.com/channels/1132231889290285117/1132602615570649110/1175108298614243328), we think this reduces readability.
For the time being, we decide not to fix this divergence.

**Comment by Conaclos** [comment]
## Divergence

Prettier adds a trailing comma to type parameters of arrow functions.

Input

```ts
<const T>() => {};
```

Diff

```diff
-<const T,>() => {};
+<const T>() => {};
```

[Playground link](https://biomejs.dev/playground?code=PABjAG8AbgBzAHQAIABUAD4AKAApACAAPQA%2BACAAewB9ADsA)

Prettier Tests

- [typescript/arrows/type_params.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptarrowstype_paramsts)
- [typescript/compiler/contextualSignatureInstantiation2.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescriptcompilercontextualsignatureinstantiation2ts)
- [typescript/typeparams/const.ts](https://github.com/biomejs/biome/blob/main/crates/biome_js_formatter/report_incompatible.md#typescripttypeparamsconstts)


## Reason

Prettier outputs an inconsistent behavior, because it adds a trailing comma to type parameter lists of arrow functions, and removes any trailing comma to type parameter lists of classes and interfaces.
To be consistent we decided to diverge and to not add this trailing comma.

**Comment by thorn0** [comment]
> Prettier encloses optional chains between parentheses

Removing the parentheses here changes the meaning of the code. The call or member expression becomes part of the optional chain.

**Comment by fisker** [comment]
> Prettier does this only in TSX to disambiguate generics from JSX elements.

That has been updated, [changelog](https://prettier.io/blog/2023/07/05/3.0.0#fix-missing-required-comma-in-type-parameters-14688httpsgithubcomprettierprettierpull14688-by-fiskerhttpsgithubcomfisker-sosukesuzukihttpsgithubcomsosukesuzuki)

**Comment by Conaclos** [comment]
> Prettier does this only in TSX to disambiguate generics from JSX elements.

> That has been updated, [changelog](https://prettier.io/blog/2023/07/05/3.0.0#fix-missing-required-comma-in-type-parameters-14688httpsgithubcomprettierprettierpull14688-by-fiskerhttpsgithubcomfisker-sosukesuzukihttpsgithubcomsosukesuzuki)

You are right. In fact, Biome already keeps the trailing comma. However, Biome seems not able to parse the `const` modifier for the first type parameter of an arrow function. This is an issue on our side. Thanks for bringing it to light :)

> Removing the parentheses here changes the meaning of the code. The call or member expression becomes part of the optional chain.

Thanks for highlighting this. We have to think about that.

**Comment by Conaclos** [comment]
## Divergence

Prettier adds a trailing comma to type parameters of arrow functions even when it is not required.

Input (`tsx`)

```tsx
<T = unknown>() => {};
```

Diff

```diff
- <T = unknown,>() => {};
+ <T = unknown>() => {};
```

## Reason

In some specific cases, a type parameter list of an arrow function requires a trailing comma to distinguish it from a JSX element.
When a default type is provided, this trailing comma is not required.
Here, we diverge from Prettier, because we think it better respects the original intent of Prettier, which was to add a trailing comma only when required.

**Comment by waynebloss** [comment]
## Divergence

Input (`ts`)

```ts
const A = {
  x: "x",
  y: "y",
  z: "z",
};

const {
  x,
  y,
  z,
} = A;
```

Diff

```diff
const A = {
  x: "x",
  y: "y",
  z: "z",
};

- const {x, y, z} = A;
+ const {
+  x,
+  y,
+  z,
+ };
```

## Reason

There was never any good reason given by Prettier why these should be treated differently.

**Comment by GabenGar** [comment]
Pretty sure the last case depends on the line length of the destructuring statement, past certain length it newlines all the keys. IMO Prettier is in the right there because newlining in all cases feels not so much different from doing `const keyName = A.keyName` for each key, for which destructuring was supposed to be a sugar in the first place. But long single-line desructuring statement can become obnoxious to read (especially when paired with nested destructuring).

**Comment by waynebloss** [comment]
@thorn0 Thanks, I didn't start a discussion. I left a Divergence and then someone responded to that, which I had no control over. Perhaps your message was meant for @GabenGar ?

**Comment by faultyserver** [comment]
This thread is specifically to list out cases where Biome has _already intentionally decided_ to deviate from Prettier's behavior and explain the rationale for that as an easily accessible source of truth for any differences that people might encounter. It's not really intended as a place to _point out_ discrepancies between Biome or Prettier or seek clarification about them.

If you'd like to _ask_ why Biome is different than Prettier in a specific case, then opening an issue about it would be the best place to start (or asking in Discord, or a Discussion). If it is then decided to _not_ be remedied by changing Biome, then a solidified, well-established answer will be added to this thread.

If you'd like to ask why _Prettier_ decides to format things in a certain way (i.e., Biome does this because Prettier does, but why does Prettier do it), then the Prettier repository would be the appropriate place to ask.

In this situation, Biome is just matching Prettier's behavior, so the question should be asked to the Prettier team, or asked in a new issue for Biome to _consider_ deviating.

**Comment by waynebloss** [comment]
> This thread is specifically to list out cases where Biome has already intentionally decided to deviate from Prettier's behavior...

Ahhh, thanks @faultyserver that was not clear.
